### PR TITLE
Update data_loader.py to Ensure Reproducibility in Multi-Process Environments with Dataloader Shuffle

### LINKED ISSUE #2574: Cant guarantee the reproducibility when using DataLoader shuffle
### System Info

```Shell
- `Accelerate` version: 0.28.0
- Platform: Windows-10-10.0.22000-SP0
- Python version: 3.11.5
- Numpy version: 1.26.4
- PyTorch version (GPU?): 2.2.1+cu118 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 15.37 GB
- GPU type: GeForce GTX 1650 Ti
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

```python
import torch
from accelerate import Accelerator

accelerator = Accelerator()
dataset = range(5)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=1, shuffle=True)
dataloader = accelerator.prepare(dataloader)

for idx, batch in enumerate(dataloader):
    print(f'Batch {idx}: {batch}')
    if idx == 2:
        accelerator.save_state('ckpt')
        print('Checkpoint saved.')
    
print('Resumed from chechpoint')
accelerator.load_state('ckpt')
dataloader = accelerator.skip_first_batches(dataloader, 3)
for idx, batch in enumerate(dataloader, 3):
    print(f'Batch {idx}: {batch}')
```
And the results were:
```
Batch 0: tensor([2], device='cuda:0')
Batch 1: tensor([3], device='cuda:0')
Batch 2: tensor([0], device='cuda:0')
Checkpoint saved.
Batch 3: tensor([1], device='cuda:0')
Batch 4: tensor([4], device='cuda:0')
Resumed from chechpoint
Batch 3: tensor([1], device='cuda:0')
Batch 4: tensor([2], device='cuda:0')
```

### Expected behavior

After saving the ckpt, the original dataloader yields samples `[1], [4]`. However, when resuming from the ckpt and skipping first 3 samples as instructed by [this page](https://huggingface.co/docs/accelerate/en/usage_guides/checkpoint), the recovered dataloader yields `[1], [2]`. 

This indicates that the accelerator does not properly record and recover randomness, which may lead to the model being trained on identical samples within a single epoch. As a result, some samples may never be encountered during the entire epoch if checkpoints are set in the middle of an epoch.

### PR DESCRIPTION
**Pull request by universuen**
# What does this PR do?

This pull request addresses an oversight in the current implementation of `prepare_data_loader` function, related to the assignment of `SeedableRandomSampler` within nested sampler structures. Specifically, when employing `BatchSamplerShard`, it is possible for its encapsulated `batch_sampler` to possess its own sampler. The existing codebase does not account for this potential nesting, failing to apply `SeedableRandomSampler` to the innermost sampler when it is of type `RandomSampler`. This adjustment ensures consistent and expected behavior across all sampler layers, promoting enhanced control and predictability in sample selection processes.

Fixes #2574

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. **The 7th comment of [this page](https://github.com/huggingface/accelerate/issues/2574#issuecomment-2015671786).**
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?
@muellerzr 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by universuen** [comment]
Hi @muellerzr ! There seems to be a problem in quality check. I've run the same command on my server and it worked fine. Could you help identify the problem? Thank you!
```
make quality
```
```
ruff check .
ruff format --check .
126 files already formatted
doc-builder style src/accelerate docs/source --max_len 119 --check_only
```

**Comment by muellerzr** [comment]
Be sure you have the right versions installed by doing `pip install -e .[quality]`

(I’ll give this a review after the weekend :) )

**Comment by universuen** [comment]
> Be sure you have the right versions installed by doing `pip install -e .[quality]`
> 
> (I’ll give this a review after the weekend :) )
> 
> also be sure to do both `make style; make quality`

Oh! Apologies for not realizing it's the weekend. I've executed these cmds and didn't observe any abnormalities. Here is the log:
```
pip install -e .[quality]
```
```
Obtaining file:///storage_fast/user/hurricane/accelerate
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: numpy>=1.17 in /storage/user/miniconda3/lib/python3.12/site-packages (1.26.3)
Requirement already satisfied: packaging>=20.0 in /storage/user/miniconda3/lib/python3.12/site-packages (23.1)
Requirement already satisfied: psutil in /storage/user/miniconda3/lib/python3.12/site-packages (5.9.8)
Requirement already satisfied: pyyaml in /storage/user/miniconda3/lib/python3.12/site-packages (6.0.1)
Requirement already satisfied: torch>=1.10.0 in /storage/user/miniconda3/lib/python3.12/site-packages (2.2.1+cu118)
Requirement already satisfied: huggingface-hub in /storage/user/miniconda3/lib/python3.12/site-packages (0.21.4)
Requirement already satisfied: safetensors>=0.3.1 in /storage/user/miniconda3/lib/python3.12/site-packages (0.4.2)
Requirement already satisfied: black~=23.1 in /storage/user/miniconda3/lib/python3.12/site-packages (23.12.1)
Requirement already satisfied: hf-doc-builder>=0.3.0 in /storage/user/miniconda3/lib/python3.12/site-packages (0.5.0)
Requirement already satisfied: ruff~=0.2.1 in /storage/user/miniconda3/lib/python3.12/site-packages (0.2.2)
Requirement already satisfied: click>=8.0.0 in /storage/user/miniconda3/lib/python3.12/site-packages (from black~=23.1) (8.1.7)
Requirement already satisfied: mypy-extensions>=0.4.3 in /storage/user/miniconda3/lib/python3.12/site-packages (from black~=23.1) (1.0.0)
Requirement already satisfied: pathspec>=0.9.0 in /storage/user/miniconda3/lib/python3.12/site-packages (from black~=23.1) (0.12.1)
Requirement already satisfied: platformdirs>=2 in /storage/user/miniconda3/lib/python3.12/site-packages (from black~=23.1) (3.10.0)
Requirement already satisfied: GitPython in /storage/user/miniconda3/lib/python3.12/site-packages (from hf-doc-builder>=0.3.0) (3.1.42)
Requirement already satisfied: tqdm in /storage/user/miniconda3/lib/python3.12/site-packages (from hf-doc-builder>=0.3.0) (4.65.0)
Requirement already satisfied: nbformat in /storage/user/miniconda3/lib/python3.12/site-packages (from hf-doc-builder>=0.3.0) (5.10.3)
Requirement already satisfied: filelock in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (3.9.0)
Requirement already satisfied: typing-extensions>=4.8.0 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (4.8.0)
Requirement already satisfied: sympy in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (1.12)
Requirement already satisfied: networkx in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (3.2.1)
Requirement already satisfied: jinja2 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (3.1.2)
Requirement already satisfied: fsspec in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (2024.2.0)
Requirement already satisfied: nvidia-cuda-nvrtc-cu11==11.8.89 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.8.89)
Requirement already satisfied: nvidia-cuda-runtime-cu11==11.8.89 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.8.89)
Requirement already satisfied: nvidia-cuda-cupti-cu11==11.8.87 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.8.87)
Requirement already satisfied: nvidia-cudnn-cu11==8.7.0.84 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (8.7.0.84)
Requirement already satisfied: nvidia-cublas-cu11==11.11.3.6 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.11.3.6)
Requirement already satisfied: nvidia-cufft-cu11==10.9.0.58 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (10.9.0.58)
Requirement already satisfied: nvidia-curand-cu11==10.3.0.86 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (10.3.0.86)
Requirement already satisfied: nvidia-cusolver-cu11==11.4.1.48 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.4.1.48)
Requirement already satisfied: nvidia-cusparse-cu11==11.7.5.86 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.7.5.86)
Requirement already satisfied: nvidia-nccl-cu11==2.19.3 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (2.19.3)
Requirement already satisfied: nvidia-nvtx-cu11==11.8.86 in /storage/user/miniconda3/lib/python3.12/site-packages (from torch>=1.10.0) (11.8.86)
Requirement already satisfied: requests in /storage/user/miniconda3/lib/python3.12/site-packages (from huggingface-hub) (2.31.0)
Requirement already satisfied: gitdb<5,>=4.0.1 in /storage/user/miniconda3/lib/python3.12/site-packages (from GitPython->hf-doc-builder>=0.3.0) (4.0.11)
Requirement already satisfied: MarkupSafe>=2.0 in /storage/user/miniconda3/lib/python3.12/site-packages (from jinja2->torch>=1.10.0) (2.1.3)
Requirement already satisfied: fastjsonschema in /storage/user/miniconda3/lib/python3.12/site-packages (from nbformat->hf-doc-builder>=0.3.0) (2.19.1)
Requirement already satisfied: jsonschema>=2.6 in /storage/user/miniconda3/lib/python3.12/site-packages (from nbformat->hf-doc-builder>=0.3.0) (4.21.1)
Requirement already satisfied: jupyter-core in /storage/user/miniconda3/lib/python3.12/site-packages (from nbformat->hf-doc-builder>=0.3.0) (5.7.2)
Requirement already satisfied: traitlets>=5.1 in /storage/user/miniconda3/lib/python3.12/site-packages (from nbformat->hf-doc-builder>=0.3.0) (5.14.2)
Requirement already satisfied: charset-normalizer<4,>=2 in /storage/user/miniconda3/lib/python3.12/site-packages (from requests->huggingface-hub) (2.0.4)
Requirement already satisfied: idna<4,>=2.5 in /storage/user/miniconda3/lib/python3.12/site-packages (from requests->huggingface-hub) (3.4)
Requirement already satisfied: urllib3<3,>=1.21.1 in /storage/user/miniconda3/lib/python3.12/site-packages (from requests->huggingface-hub) (2.1.0)
Requirement already satisfied: certifi>=2017.4.17 in /storage/user/miniconda3/lib/python3.12/site-packages (from requests->huggingface-hub) (2024.2.2)
Requirement already satisfied: mpmath>=0.19 in /storage/user/miniconda3/lib/python3.12/site-packages (from sympy->torch>=1.10.0) (1.3.0)
Requirement already satisfied: smmap<6,>=3.0.1 in /storage/user/miniconda3/lib/python3.12/site-packages (from gitdb<5,>=4.0.1->GitPython->hf-doc-builder>=0.3.0) (5.0.1)
Requirement already satisfied: attrs>=22.2.0 in /storage/user/miniconda3/lib/python3.12/site-packages (from jsonschema>=2.6->nbformat->hf-doc-builder>=0.3.0) (23.2.0)
Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /storage/user/miniconda3/lib/python3.12/site-packages (from jsonschema>=2.6->nbformat->hf-doc-builder>=0.3.0) (2023.12.1)
Requirement already satisfied: referencing>=0.28.4 in /storage/user/miniconda3/lib/python3.12/site-packages (from jsonschema>=2.6->nbformat->hf-doc-builder>=0.3.0) (0.34.0)
Requirement already satisfied: rpds-py>=0.7.1 in /storage/user/miniconda3/lib/python3.12/site-packages (from jsonschema>=2.6->nbformat->hf-doc-builder>=0.3.0) (0.18.0)
Checking if build backend supports build_editable ... done
Building wheels for collected packages: accelerate
  Building editable for accelerate (pyproject.toml) ... done
  Created wheel for accelerate: filename=accelerate-0.29.0.dev0-0.editable-py3-none-any.whl size=12128 sha256=cee4b8816a0be204617fa17494c7067dc3cee5f9ffa015dc45dc5272afe403d0
  Stored in directory: /tmp/pip-ephem-wheel-cache-zq385och/wheels/52/04/f9/8d449e0b508712291b9117bad0c6cdfa25f639c765b9aa730f
Successfully built accelerate
Installing collected packages: accelerate
  Attempting uninstall: accelerate
    Found existing installation: accelerate 0.29.0.dev0
    Uninstalling accelerate-0.29.0.dev0:
      Successfully uninstalled accelerate-0.29.0.dev0
Successfully installed accelerate-0.29.0.dev0
```
```
make style; make quality
```
```
ruff check . --fix
ruff format .
1 file reformatted, 125 files left unchanged
doc-builder style src/accelerate docs/source --max_len 119
ruff check .
ruff format --check .
126 files already formatted
doc-builder style src/accelerate docs/source --max_len 119 --check_only
```

**Comment by muellerzr** [comment]
Well it says right there that one file was reformatted :)

**Comment by universuen** [comment]
> Well it says right there that one file was reformatted :)

Ah I finally got it. Sorry I am not familiar with this `ruff` tool. The newly formatted code was committed.

**Comment by universuen** [comment]
For your convenience, I've prepared this POC code:
```python
import torch
from accelerate import Accelerator, DataLoaderConfiguration
from accelerate.utils import set_seed
from accelerate import notebook_launcher


def main():
    set_seed(42)
    
    accelerator = Accelerator(
        dataloader_config=DataLoaderConfiguration(use_seedable_sampler=True)
    )
    dataset = range(10)
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=1, shuffle=True)
    dataloader = accelerator.prepare(dataloader)
    print(type(dataloader.batch_sampler.batch_sampler.sampler))

notebook_launcher(main, num_processes=2, use_port='8000')
```
```
<class 'torch.utils.data.sampler.RandomSampler'>
<class 'torch.utils.data.sampler.RandomSampler'>
```

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for providing this patch, nice investigation of the issue. Would it be possible to add a unit test, perhaps based on your provided example?

**Comment by universuen** [comment]
> Thanks for providing this patch, nice investigation of the issue. Would it be possible to add a unit test, perhaps based on your provided example?

Hi! Thank you for your prompt response. I've incorporated a unit test as suggested; however, I'm not sure whether this addition aligns with the expected formal coding standards.

File: tests/test_data_loader.py [hunk 1]
```diff
@@ -396,3 +398,15 @@ def test_end_of_dataloader_dispatcher(self):
         # Test it also works on the second iteration
         for idx, _ in enumerate(dataloader):
             assert dataloader.end_of_dataloader == (idx == 3)
+
+    def test_seedable_random_sampler_is_fully_applied(self):
+        dataloader = DataLoader(list(range(10)), batch_size=1, shuffle=True)
```
Comment (BenjaminBossan) [inline tests/test_data_loader.py:403]: Could you please instantiate the accelerator by calling `Accelerator()` at the top of the function body, same as in the other tests?

Comment (universuen) [reply tests/test_data_loader.py:404]: Okay! I've add the initialization like `DataLoaderTester.test_end_of_dataloader_dispatcher` does. Should I pass `DataLoaderConfiguration` to `Accelerator` and call its `prepare()` method to do the test?

File: tests/test_data_loader.py [hunk 2]
```diff
@@ -396,3 +398,16 @@ def test_end_of_dataloader_dispatcher(self):
...
+        prepared_data_loader = prepare_data_loader(
+            dataloader=dataloader,
+            num_processes=2,
+            process_index=0,
+            use_seedable_sampler=True,
+        )
+        batch_sampler = prepared_data_loader.batch_sampler
+        assert isinstance(batch_sampler, BatchSamplerShard)
+        assert isinstance(batch_sampler.batch_sampler.sampler, SeedableRandomSampler)
```
Comment (muellerzr) [inline tests/test_data_loader.py:413]: This should generally be under `src/accelerate/test_utils/scripts/test_script.py`, you'll find similar tests in there for `prepare_data_loader` you can mimic.

Let's move that there please?

**Review by muellerzr — COMMENTED** [review]
Thanks, overall this looks okay to me. Let's change where this test is though to align with how we run multi-device tests please :)

File: tests/test_data_loader.py (hunk 2 above)
Comment (universuen) [reply tests/test_data_loader.py:413]: Yeah of course! I'll move it now.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for the fix, looks good on my end.

**Review by muellerzr — APPROVED** [review]
Looks good on my end as well! Let's keep an eye on `main` to see if a red X shows up from our multi-GPU tests
